include: Implement API's for cache flush and cache invalidate

aasthagr opened the pull request:
include: Implement API's for cache flush and cache invalidate
arch: arc: core: Add Implementation function & prototype for arc

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

aasthagr committed:
include: Implement API's for cache flush and cache invalidate

arch: arc: core: Add Cache Implementation function & prototype for arc

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

andrewboie commented on arch/arc/core/cache.c:
```diff
@@ -135,12 +135,44 @@ static void dcache_flush_mlines(u32_t start_addr, u32_t size)
  * @return N/A
```
Move this doxygen to the dcache_flush_mlines function

aasthagr replied:
This doxygen is already exists for dcache_flush_mlines function. Do you want me to remove this from arch_cache_flush?

andrewboie replied:
This is what I'd like to see:
1) The doxygen for the arch functions should always be in the header where the prototype is declared.
2) Any duplicate doxygen at the site of arch function implementations should be removed. https://en.wikipedia.org/wiki/Don%27t_repeat_yourself
3) HOWEVER: any existing comments specific to the implementation of an arch function should be preserved, but not in doxygen format, just a comment.

andrewboie commented on include/cache.h:
```diff
@@ -13,48 +13,43 @@
-#if defined(CONFIG_ARCH_CACHE_FLUSH_DETECT)
-	typedef _sys_cache_flush_sig(_sys_cache_flush_t);
-	extern _sys_cache_flush_t *sys_cache_flush;
+static inline void sys_cache_flush(void *addr, size_t size, u32_t flags)
```
These APIs are part of the Zephyr public interface and will need doxygen documentation.

andrewboie commented on include/cache.h:
```diff
@@ -13,48 +13,43 @@
- */
-
-static inline _sys_cache_flush_sig(sys_cache_flush)
+static inline void sys_cache_invd(void *addr, size_t size, u32_t flags)
```
Same, needs documentation.

andrewboie commented on include/cache.h:
```diff
@@ -13,48 +13,43 @@
-	extern size_t sys_cache_line_size;
-#elif defined(CONFIG_CACHE_LINE_SIZE)
-	#define sys_cache_line_size CONFIG_CACHE_LINE_SIZE
+static inline size_t sys_cache_line_size_get(void)
```
Need doxygen documentation.

andrewboie commented on include/sys/arch_interface.h:
```diff
@@ -675,6 +675,21 @@ extern u64_t arch_timing_value_swap_temp;
+ */
+
+#ifdef CONFIG_CACHE_FLUSHING
+void arch_cache_flush(void *addr, size_t size, u32_t flags);
```
This will also need documentation. Can be brief, with a "@see" reference to the APIs in cache.h.

tyrantlizard77 reviewed (changes requested):
Please fix all the errors found by the CI checks.

Don't we need to update the x86 cache code, since it was following the old interface?

andrewboie commented:
I think we will also need a patch against libmetal:

```
FAILED: modules/libmetal/libmetal/lib/CMakeFiles/metal.dir/device.c.obj 
ccache /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBUILD_VERSION=zephyr-v2.2.0-2052-g40b6cc1591ac -DCORE_CM4 -DHSE_VALUE=8000000 -DKERNEL -DMETAL_INTERNAL -DSTM32MP157Cxx -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR__=1 -Imodules/libmetal/libmetal/lib/include -I/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include -Izephyr/include/generated -I/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/soc/arm/st_stm32/stm32mp1 -I/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/lib/open-amp/. -I/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers -I/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/testsuite/include -I/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/testsuite/ztest/include -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/cmsis/CMSIS/Core/Include -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/stm32/stm32cube/stm32mp1xx/soc -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/stm32/stm32cube/stm32mp1xx/drivers/include -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/stm32/stm32cube/stm32mp1xx/drivers/include/Legacy -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/lib/open-amp/open-amp/lib/include -isystem /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/lib/libc/minimal/include -isystem /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include -isystem /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include-fixed -Os -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out/stm32mp157c_dk2/tests/kernel/common/kernel.common/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -mcpu=cortex-m4 -mthumb -mabi=aapcs -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-address-of-packed-member -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/kernel/common=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT modules/libmetal/libmetal/lib/CMakeFiles/metal.dir/device.c.obj -MF modules/libmetal/libmetal/lib/CMakeFiles/metal.dir/device.c.obj.d -o modules/libmetal/libmetal/lib/CMakeFiles/metal.dir/device.c.obj   -c /home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/libmetal/libmetal/lib/device.c
In file included from modules/libmetal/libmetal/lib/include/metal/cache.h:15,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/libmetal/libmetal/lib/device.c:16:
modules/libmetal/libmetal/lib/include/metal/system/zephyr/cache.h: In function '__metal_cache_flush':
modules/libmetal/libmetal/lib/include/metal/system/zephyr/cache.h:28:18: error: passing argument 1 of 'sys_cache_flush' makes pointer from integer without a cast [-Werror=int-conversion]
   28 |  sys_cache_flush((vaddr_t) addr, len);
      |                  ^~~~~~~~~~~~~~
      |                  |
      |                  unsigned int
In file included from modules/libmetal/libmetal/lib/include/metal/system/zephyr/cache.h:19,
                 from modules/libmetal/libmetal/lib/include/metal/cache.h:15,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/libmetal/libmetal/lib/device.c:16:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/cache.h:16:42: note: expected 'void *' but argument is of type 'unsigned int'
   16 | static inline void sys_cache_flush(void *addr, size_t size, u32_t flags)
      |                                    ~~~~~~^~~~
In file included from modules/libmetal/libmetal/lib/include/metal/cache.h:15,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/libmetal/libmetal/lib/device.c:16:
modules/libmetal/libmetal/lib/include/metal/system/zephyr/cache.h:28:2: error: too few arguments to function 'sys_cache_flush'
   28 |  sys_cache_flush((vaddr_t) addr, len);
      |  ^~~~~~~~~~~~~~~
In file included from modules/libmetal/libmetal/lib/include/metal/system/zephyr/cache.h:19,
                 from modules/libmetal/libmetal/lib/include/metal/cache.h:15,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/libmetal/libmetal/lib/device.c:16:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/cache.h:16:20: note: declared here
   16 | static inline void sys_cache_flush(void *addr, size_t size, u32_t flags)
      |                    ^~~~~~~~~~~~~~~
```

I'm not sure how we synchronize changes between two repos like this.

andrewboie commented on include/cache.h:
```diff
@@ -13,45 +13,75 @@
- * Provide NOP APIs for systems that do not have caches.
+/**
+ *
+ * @brief Invalidate d-cache lines
```
I think some supporting comments should be here, like has been done with `sys_cache_flush()`

andrewboie commented on include/cache.h:
```diff
@@ -13,45 +13,75 @@
+ *
+ * The API is provided to get the cache line size.
+ *
+ * @return size of the cache line.
```
"...or 0 if the cache is not enabled"

andrewboie commented on arch/arc/core/cache.c:
```diff
@@ -117,29 +100,38 @@ static void dcache_flush_mlines(u32_t start_addr, u32_t size)
-void sys_cache_flush(vaddr_t start_addr, size_t size)
+void arch_cache_flush(void *start_addr, size_t size, u32_t flags)
 {
+	ARG_UNUSED(flags);
```
is anything else using `dcache_flush_mlines()`? could fold its body into this function if not

aasthagr replied:
Doesn't seem like any other function is using dcache_flush_mlines, will merge it with arch_cache_flush then. Thanks

andrewboie commented:
I thought about the libmetal failure some more.

I don't think we have a clear picture right now what `flags` ought to be, let's just drop the `flags` arg for now.

For compatibility, let's `typedef void * vaddr_t` in the header. That should allow libmetal to compile.

Still need to adapt the x86 code to this.

aasthagr commented:
> I thought about the libmetal failure some more.
> 
> I don't think we have a clear picture right now what `flags` ought to be, let's just drop the `flags` arg for now.
> 
> For compatibility, let's `typedef void * vaddr_t` in the header. That should allow libmetal to compile.
> 
> Still need to adapt the x86 code to this.

I am not able to reproduce this error of libmetal & if I try to do typedef void *vaddr_t in include/cache.h file, there is an error that vaddr_t have been previously declared. Any suggestions?

andrewboie commented:
> I am not able to reproduce this error of libmetal 

I would figure this out. Look at the CI logs, find out which test/platform this is failing on.

> I try to do typedef void *vaddr_t in include/cache.h file, there is an error that vaddr_t have been previously declared. Any suggestions?

Yes, find the conflicting definition of `vaddr_t` and update it. `grep` or tools like https://github.com/ggreer/the_silver_searcher are your friend here. I found the definition in `include/arch/common/addr_types.h.`. I would suggest `void *` for `vaddr_t` and `uintptr_t` for `paddr_t`.

ruuddw commented:
Great to see the cache APIs get addressed. There's actually a very old issue on this: #2647.
Should support for instruction caches be added as well? As a minimum i-cache invalidate would be good, though the use-cases for it are not yet supported in Zephyr (dynamic loading/overlays). 
The sys_cache_xxx names now don't reflect i or d. Add a 'd'?

andrewboie commented:
> Should support for instruction caches be added as well? As a minimum i-cache invalidate would be good, though the use-cases for it are not yet supported in Zephyr (dynamic loading/overlays).

This isn't in scope for current efforts but could be added later. We are keeping the names the same as to not break existing code. icache invalidation will be a different, new API set.

ruuddw commented:
ok - I thought since the APIs are internal, now would be the right time to introduce consistent names. But if they are already widely used, let's stick to the current ones.

ceolin commented on arch/x86/core/ia32/cache.c:
```diff
@@ -47,66 +47,18 @@ extern size_t z_cache_line_size_get(void);
 
-	size = ROUND_UP(size, sys_cache_line_size);
-	end = virt + size;
+//	size = ROUND_UP(size, z_cache_line_size_get());
```
why is this commented ? should you ensure that size is aligned ?

ceolin commented on include/arch/common/addr_types.h:
```diff
@@ -10,8 +10,8 @@
-typedef unsigned int paddr_t;
-typedef unsigned int vaddr_t;
+typedef uintptr_t paddr_t;
+typedef void *vaddr_t;
```
why they are different ? can't you define both `uintptr_t` ?

andrewboie replied:
I asked her to do it this way. I think it is better to store virtual addresses as `void *` since they represent actual addressable memory locations, and `uintptr_t` for physical memory addresses since they should not be directly dereferenced without a conversion to physical address first. I think there is value in not having the same data types for both as it will be less easy to accidentally confuse the two.

Abhishek-brcm replied:
> `uintptr_t` for physical memory addresses since they should not be directly dereferenced without a conversion to physical address first.

Did you mean, without a conversion to virtual address first?

> The sys_cache_xxx names now don't reflect i or d. Add a 'd'?

I think we can use `sys_dcache_flush`, `sys_dcache_clean` in this PR to reflect that these are dcache operations(as already suggested by @ruuddw above). Also `sys_dcache_clean` should be considered to be added- if not here, in separate PR. icache APIs can be considered separately anyway.

andrewboie replied:
> > `uintptr_t` for physical memory addresses since they should not be directly dereferenced without a conversion to physical address first.
> 
> Did you mean, without a conversion to virtual address first?

Oops, yes that is what I meant.

ceolin commented:
overall looks good to me, just check my comments.

andyross commented on arch/arc/core/cache.c:
```diff
@@ -93,8 +76,8 @@ static void dcache_flush_mlines(u32_t start_addr, u32_t size)
-	end_addr = start_addr + size - 1;
-	start_addr &= (u32_t)(~(DCACHE_LINE_SIZE - 1));
+	end_addr = (u32_t)start_addr + size - 1;
+	*((u32_t *)start_addr) &= (u32_t)(~(DCACHE_LINE_SIZE - 1));
```
This is writing *through* the pointer and storing data.  That can't be what you intend, can it?  The original code there was just aligning it downward to the next lower cache line.

andrewboie replied:
yes this isn't right, the purpose of that code is to align the address pointer

aasthagr replied:
My bad. Addressed this issue now. Thanks for pointing that out.

andyross reviewed (changes requested):
I'm either really confused, or there's a terrible data-killing typo on arc.  :)

Organization and API looks good to me.  We might want to consider adding a "sys_cache_flush_and_invalidate()" API, as many/most architectures support that via a single mechanism and it's often a needed optimization for synchronization work.

tyrantlizard77 reviewed (changes requested):
please double check the code which is aligning start addresses, writing to memory is not what you want

ceolin commented:
> please double check the code which is aligning start addresses, writing to memory is not what you want

fair enough :)

ceolin commented on arch/x86/core/ia32/cache.c:
```diff
@@ -47,66 +47,18 @@ extern size_t z_cache_line_size_get(void);
 
-	for (; virt < end; virt += sys_cache_line_size) {
-		__asm__ volatile("clflush %0;\n\t" :  : "m"(virt));
+	for (; (u32_t)start_addr < end; (u32_t)start_addr += z_cache_line_size_get()) {
```
here you're a change the old behaviour. You probably should have to handle the case where `CONFIG_CACHE_LINE_SIZE_DETECT` is not defined.

ceolin commented on arch/x86/core/ia32/cache.c:
```diff
@@ -47,66 +47,18 @@ extern size_t z_cache_line_size_get(void);
-	return 0;
-}
-
-SYS_INIT(init_cache, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
I don't think you should remove it, this was flushing the cache during initialization.

aasthagr replied:
Yes, this should not be removed. Added this back. Thanks

ceolin reviewed (changes requested):
There are things to look in X86 changes

andyross commented on arch/arc/core/cache.c:
```diff
@@ -110,36 +94,37 @@ static void dcache_flush_mlines(u32_t start_addr, u32_t size)
 		} while (1);
-		start_addr += DCACHE_LINE_SIZE;
-	} while (start_addr <= end_addr);
+		*((u32_t *)start_addr) += DCACHE_LINE_SIZE;
```
This is still scribbling over the memory area we're trying to flush.

aasthagr replied:
Fixed this.

andyross commented on arch/arc/core/cache.c:
```diff
@@ -67,24 +67,7 @@ static void dcache_enable(void)
- *
- * @return N/A
- */
-static void dcache_flush_mlines(u32_t start_addr, u32_t size)
```
FWIW: the implementation of the new arch_cache_flush() and the old dcache_flush_mlines() look 100% identical to me (modulo that pointer dereference bug called out elsewhere).   It might be simpler from a review standpoint to limit the patch to a trivial rename (and doxygen comment removal).

aasthagr replied:
I can open a separate PR for just the rename & doxygen removal of dcache_flush_mlines() function.

andyross reviewed (approved):
I'm out of complaints.  Looks good to me.

stephanosio commented on arch/arc/core/cache.c:
```diff
@@ -67,24 +67,7 @@ static void dcache_enable(void)
-static void dcache_flush_mlines(u32_t start_addr, u32_t size)
+void arch_cache_flush(void *start_addr, size_t size)
 {
 	u32_t end_addr;
```
Should use `uintptr_t`.

p.s. I understand that we only support 32-bit ARC at the moment, but using `uintptr_t` here would ensure that no code change is necessary when 64-bit ARC support is ever added.

aasthagr replied:
Agreed. Changed it to uintptr_t type.

stephanosio commented on arch/arc/core/cache.c:
```diff
@@ -93,8 +76,9 @@ static void dcache_flush_mlines(u32_t start_addr, u32_t size)
 
-	end_addr = start_addr + size - 1;
-	start_addr &= (u32_t)(~(DCACHE_LINE_SIZE - 1));
+	end_addr = (u32_t)start_addr + size - 1;
```
same here: `uintptr_t`

stephanosio commented on arch/x86/core/ia32/cache.c:
```diff
@@ -47,15 +47,15 @@ extern size_t z_cache_line_size_get(void);
+void arch_cache_flush(void *start_addr, size_t size)
 {
-	int end;
+	u32_t end;
```
Deprecated Zephyr type; please use C99 types.

stephanosio commented on include/cache.h:
```diff
@@ -13,45 +13,78 @@
+static inline void sys_cache_flush(void *addr, size_t size)
+{
+#ifdef CONFIG_CACHE_FLUSHING
+	arch_cache_flush(addr, size);
 #else
-	extern _sys_cache_flush_sig(sys_cache_flush);
+	ARG_UNUSED(addr);
+	ARG_UNUSED(size);
 #endif
```
Better to do the following -- that way we don't need these superficial `ARG_UNUSED`s
```
if (IS_ENABLED(CONFIG_CACHE_FLUSHING)) {
    ...
}
```

aasthagr replied:
Instead of IS_ENABLED , I am just using ifdef CONFIG_CACHE_FLUSHING now and removed the else part where there were ARG_UNUSED statements. Let me know if this is fine. Thanks

stephanosio replied:
The problem with removing `ARG_UNUSED` is that, as far as C compiler (not preprocessor) is concerned, the `addr` and `size` are unused when `CONFIG_CACHE_FLUSHING=n` and it will display warnings.

Using `if (IS_ENABLED ...` instead of preprocessor ifdef ensures that C compiler is aware of its usage. This is, in fact, a recommended coding practice in the Linux kernel.

aasthagr replied:
Changed it to IS_ENABLED and added the declaration for functions arch_dcache_flush and arch_dcache_invd. Thanks

stephanosio reviewed (commented):
I suppose this is better than what we had before; though, a complete cache management API would require multi-level cache management with scoping for SMP systems.

andrewboie commented on include/cache.h:
```diff
@@ -13,45 +13,78 @@
+ *
+ * @return N/A
+ */
+static inline void sys_cache_flush(void *addr, size_t size)
```
@ruuddw and @Abhishek-brcm have a good point; if we later add icache management, makes more sense to call these APIs `sys_dcache_flush()`, `sys_dcache_invd()`, etc. Please rename, along with the associated arch APIs.

tyrantlizard77 reviewed (changes requested):
Getting close. Agree with the naming suggestion. For RAM addresses please always use `void *` or `uintptr_t`, and `size_t` for region sizes, and not `u32_t`/`uint32_t`

andrewboie commented on arch/arc/core/cache.c:
```diff
@@ -67,34 +67,18 @@ static void dcache_enable(void)
-	end_addr = start_addr + size - 1;
-	start_addr &= (u32_t)(~(DCACHE_LINE_SIZE - 1));
+	end_addr = (uintptr_t)start_addr + size;
+	start_addr = (void *)(((uintptr_t)start_addr) &
```
consider using `ROUND_DOWN(start_addr, DCACHE_LINE_SIZE)` for clarity. This macro and its companion ROUND_UP are found in include/sys/util.h

andrewboie commented on arch/arc/core/cache.c:
```diff
@@ -110,36 +94,37 @@ static void dcache_flush_mlines(u32_t start_addr, u32_t size)
+		return;
+	}
+	end_addr = (uintptr_t)start_addr + size;
+	start_addr = (void *)(((uintptr_t)start_addr) &
```
also consider ROUND_DOWN() here

aasthagr replied:
Resolved.

andrewboie commented on arch/arc/core/cache.c:
```diff
@@ -110,36 +94,37 @@ static void dcache_flush_mlines(u32_t start_addr, u32_t size)
+		printk("Warn: cache not available or %d size\n", size);
+		return;
+	}
+	end_addr = (uintptr_t)start_addr + size;
```
style: consider a local uintptr_t variable for the start address, rather than casting the arg in lots of places.

aasthagr replied:
Addressed with local variable start_addr_temp. thanks

tyrantlizard77 reviewed (commented):
@aasthagr some minor comments, can you please rebase this on the latest master? this is against an older version of the codebase and there are conflicts

Abhishek-brcm commented on include/cache.h:
```diff
@@ -13,45 +13,72 @@
- * architecture-agnostic manner. The functions do nothing. The cache line size
- * value is always 0.
+ * No alignment is required for either addr or size, but since
+ * sys_cache_invd() iterates on the d-cache lines, a d-cache line alignment for
```
comment(s) need to updated in accordance with API name.

aasthagr replied:
Updated the comments and rebased. Thanks.

Abhishek-brcm reviewed (commented):
Minor comment. Please rebase.

aasthagr committed:
arch: x86: core: Add cache flush function for x86

Adding just the cache flush function for x86. The name
arch_cache_flush comply with API names in include/cache.h

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

aasthagr commented on arch/arc/core/cache.c:
```diff
@@ -93,8 +76,9 @@ static void dcache_flush_mlines(u32_t start_addr, u32_t size)
 
-	end_addr = start_addr + size - 1;
-	start_addr &= (u32_t)(~(DCACHE_LINE_SIZE - 1));
+	end_addr = (u32_t)start_addr + size - 1;
```
Done.

stephanosio commented on include/cache.h:
```diff
@@ -13,45 +13,75 @@
+ */
+static inline void sys_dcache_invd(void *addr, size_t size)
+{
+if (IS_ENABLED(CONFIG_CACHE_FLUSHING)) {
+    arch_dcache_invd(addr, size);
+}
```
Nits: indentation (also need to be TAB as with the rest in this file)

stephanosio commented on include/cache.h:
```diff
@@ -13,45 +13,75 @@
-	ARG_UNUSED(size);
-
-	/* do nothing */
+if (IS_ENABLED(CONFIG_CACHE_FLUSHING)) {
+    arch_dcache_flush(addr, size);
+}
```
same deal here

stephanosio commented on arch/arc/core/cache.c:
```diff
@@ -67,34 +67,19 @@ static void dcache_enable(void)
- */
-static void dcache_flush_mlines(uint32_t start_addr, uint32_t size)
+void arch_dcache_flush(void *start_addr, size_t size)
 {
-	uint32_t end_addr;
+	uintptr_t start_addr_temp = (uintptr_t)start_addr;
+	uintptr_t end_addr;
```
Just a suggestion: it would be more straight forward if we rename `void *start_addr` to `start_addr_ptr` and `uintptr_t start_addr_temp` to simply `start_addr`.

aasthagr replied:
Renaming start_addr  to start_addr_ptr  is okay. But we needed a local variable start_addr_temp  to typecast the void* start_addr to uintptr_t . So that typecasting is not done everywhere with start_addr. Let me know if you have any suggestions here given we would need a local variable whether it be start_addr_temp or any other name. Thanks

Abhishek-brcm replied:
I think @stephanosio is suggesting:
```
void arch_dcache_flush(void *start_addr_ptr, size_t size)
{
uintptr_t start_addr = (uintptr_t)start_addr_ptr;
...
Now you can use start_addr everywhere in the function instead of start_addr_temp
...
}
```

Abhishek-brcm commented:
Looks like API raname is causing CI build failures... not sure what libmetal is...

aasthagr commented:
> Looks like API rename is causing CI build failures... not sure what libmetal is... please revert back to original names for now..

Yeahh, changed back to the original names.

Abhishek-brcm reviewed (approved):
Would like to move forward with this PR. Improvements can be taken up later.

Abhishek-brcm commented on arch/arc/core/cache.c:
```diff
@@ -110,36 +95,37 @@ static void dcache_flush_mlines(uint32_t start_addr, uint32_t size)
 		} while (1);
-		start_addr += DCACHE_LINE_SIZE;
-	} while (start_addr <= end_addr);
+		start_addr_temp += DCACHE_LINE_SIZE;
```
start_addr_temp is increased by d-cache line size.  Does line 87 needs to be changed accordingly? 		`z_arc_v2_aux_reg_write(_ARC_V2_DC_FLDL, start_addr);` --> `z_arc_v2_aux_reg_write(_ARC_V2_DC_FLDL, start_addr_temp); `with current logic? [please follow @stephanosio's suggestion above, then this change shouldn't be required]

Abhishek-brcm replied:
> Would like to move forward with this PR. Improvements can be taken up later.

Have a comment to be addressed first..

Abhishek-brcm reviewed (dismissed):
Have a comment that needs to be addressed..

andrewboie commented on arch/arc/core/cache.c:
```diff
@@ -111,35 +96,36 @@ static void dcache_flush_mlines(uint32_t start_addr, uint32_t size)
- * @return N/A
- */
+	if (!dcache_available() || (size == 0U)) {
+		printk("Warn: cache not available or %d size\n", size);
```
I suggest dropping this printk() for symmetry with arch_dcache_flush() which simply returns

aasthagr replied:
Removed the printk().

andrewboie commented on arch/x86/core/ia32/cache.c:
```diff
@@ -47,15 +47,15 @@ extern size_t z_cache_line_size_get(void);
+		__asm__ volatile("clflush %0;\n\t" :  : "m"((uintptr_t)start_addr));
 	}
 
 	__asm__ volatile("mfence;\n\t");
```
there's some code below here for init_cache_flush() do we actually need it? I think it should be removed

andrewboie replied:
basically please remove unnecessary stuff from this C file.

andrewboie commented on include/cache.h:
```diff
@@ -13,45 +13,75 @@
 
-#define _sys_cache_flush_sig(x) void (x)(vaddr_t virt, size_t size)
+void arch_dcache_flush(void *addr, size_t size);
+void arch_dcache_invd(void *addr, size_t size);
```
yeah there is still cleanup work to do in the x86 file.

aasthagr replied:
Made the suggested changes in x86 file. Please review that again. Thanks

tyrantlizard77 reviewed (changes requested):
still needs some cleanup in the x86 implementation, please get rid of anything in that file that isn't necessary

ruuddw commented on arch/arc/core/cache.c:
```diff
@@ -111,35 +96,35 @@ static void dcache_flush_mlines(uint32_t start_addr, uint32_t size)
-{
-	dcache_flush_mlines((uint32_t)start_addr, (uint32_t)size);
-}
+	key = irq_lock(); /* -enter critical section- */
```
The flush routine uses arch_irq_lock (in case of SMP: local CPU interrupt disable, not a global spinlock). That would be enough here as well.
(and, there's a tradeoff between locking overhead and interrupt blocking time: the lock is only required within the loop. If small regions are flushed, this is better - let's leave it like this)

ruuddw replied:
to avoid confusion: can you change the irq_lock to arch_irq_lock (the "let's leave it like this" referred to moving the lock into the loop).

aasthagr replied:
Changed it to arch_irq_lock to be consistent as well , Thank you!!

stephanosio commented on arch/x86/core/ia32/cache.c:
```diff
@@ -47,15 +47,15 @@ extern size_t z_cache_line_size_get(void);
 
-_sys_cache_flush_sig(_cache_flush_clflush)
+void arch_dcache_flush(void *start_addr, size_t size)
 {
-	int end;
+	uintptr_t end;
```
Suggestion: shall we do the same thing we did for ARC here just to make it a bit prettier?
```
void arch_dcache_flush(void *start_addr, size_t size)
{
	uintptr_t start = (uintptr_t)start_addr;
	...
```

stephanosio reviewed (approved):
Looks ok overall

ruuddw reviewed (approved):
The ARC part looks good now. Aastha, thanks for addressing all comments.

nashif commented on include/sys/arch_interface.h:
```diff
@@ -675,6 +675,42 @@ extern uint64_t arch_timing_value_swap_temp;
+ */
+
+#ifdef CONFIG_CACHE_FLUSHING
+
```
newline and spacing are inconsistent in this file, please fix that.

nashif commented on include/sys/arch_interface.h:
```diff
@@ -675,6 +675,42 @@ extern uint64_t arch_timing_value_swap_temp;
+ * @see sys_cache_invd
+ */
+void arch_dcache_invd(void *addr, size_t size);
+#ifndef CONFIG_CACHE_LINE_SIZE
```
here to, add a new line on top and have the if just before the doxygen comment.

aasthagr replied:
Done. Please check if it is done correctly. Thanks

nashif commented on include/sys/arch_interface.h:
```diff
@@ -675,6 +675,42 @@ extern uint64_t arch_timing_value_swap_temp;
+ *
+ * @see sys_cache_line_size_get
+ */
+size_t arch_cache_line_size_get(void);
```
do we have any users/implementers of this anywhere?

aasthagr replied:
No the implementation for arch_cache_line_size_get doesnt exist as of now.

nashif reviewed (commented):
one more thing to improve is Kconfig for caching, right now it is done per architecture (arc, x86) in inconsistent ways, we should look into moving the caching kconfig up and make them work for all architectures in a consistent way because we do now have an architecture interface for this that is public API.

This can be done in a followup PR though. #26760

Also, we need to add tests for all of this.

andrewboie commented:
> Also, we need to add tests for all of this.

Agreed, but for this particular case I suggest we let this PR go in and file a bug on this (if there isn't already, we have had the sys_cache_flush function around since before I joined the project). it's a hard problem to prove in code that the CPU actually performed the requested caching operation. I can't think of a good generic way to do it, and the ways I can imagine are pretty architecture-specific.

nashif commented:
> Agreed, but for this particular case I suggest we let this PR go in and file a bug on this

right, I thought I said this above, maybe i was not clear.

Abhishek-brcm reviewed (approved):
Looks good to me as starting point to add generic cache maintenance APIs.

ceolin commented on include/cache.h:
```diff
@@ -13,45 +13,75 @@
+#else
+	return 0;
+#endif /* CONFIG_CACHE_FLUSHING */
+}
```
as @nashif pointed `arch_cache_line_size_get` is not being implemented. Is it possible one enable `CONFIG_CACHE_FLUSHING` and end up with an undefined reference. Or am I wrong and there is a contract between `CONFIG_CACHE_FLUSHING` and this API.

ceolin commented on arch/x86/core/ia32/cache.c:
```diff
@@ -47,15 +47,15 @@ extern size_t z_cache_line_size_get(void);
-	for (; virt < end; virt += sys_cache_line_size) {
-		__asm__ volatile("clflush %0;\n\t" :  : "m"(virt));
+	for (; (uintptr_t)start_addr < end; (uintptr_t)start_addr += sys_cache_line_size) {
+		__asm__ volatile("clflush %0;\n\t" :  : "m"((uintptr_t)start_addr));
```
ok, looking this file and the previous commit, shouldn't sys_cache_line_size variable be replaced with `sys_cache_line_size` and `z_cache_line_size_get` renamed to `arch_cache_line_size_get` ?

ioannisg commented on include/arch/common/addr_types.h:
```diff
@@ -10,8 +10,8 @@
 #define ZEPHYR_INCLUDE_ARCH_X86_ADDR_TYPES_H_
```
BTW, this is wrong, now, it should probably be COMMON instead of X86

ioannisg reviewed (approved):
API  looks ok to me

ceolin commented:
@aasthagr The new API looks good to me. Regarding my comment about `z_cache_line_size_get` does it make sense ? If yes are you going to address it in a further pr ?

nashif commented:
> If yes are you going to address it in a further pr ?

IMO this needs to be part of this PR, otherwise it is an incomplete implementation.

aasthagr commented:
> @aasthagr The new API looks good to me. Regarding my comment about `z_cache_line_size_get` does it make sense ? If yes are you going to address it in a further pr ?

Yes it makes sense and I am trying to figure out the options to rename z_cache_line_size_get to arch_cache_line_size_get, but there is not implementation of arch_line_size_get yet. I am thinking if using the 'CACHE_LINE_SIZE_DETECT' would be a good idea instead of implementing the arch_line_size_get as this option is enabled by default at runtime. Suggestions?

aasthagr commented:
> IMO this needs to be part of this PR, otherwise it is an incomplete implementation.

Addressed this and implemented the arch_cache_line_size_get() in cache.c for arc and x86 both. Please review the change. Thanks.
